typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+
+// Implemented on any type that can be represented as an ArbitraryData in a3-...
+trait CommandArgs { ... }
+
+fn command<Args: CommandArgs, Success: CommandReturn, Failure: CommandReturn>(
+    driver_num: u32,
+    command_num: u32,
+    args: Args) -> Result<Result<Success, Failure>, ErrorCode> { ... }
+```
+
+For libtock-c and other languages with less powerful generics, we would generate
+the interface code for each syscall driver.
+
+It is expected that userspace libraries will provide a way to register upcalls
+that have a function signature matching the upcall's concrete type instead of
+`ArbitraryData`; they'll wrap that function in one that decodes the
+`ArbitraryData`. The userspace library will pass the latter function to
+Subscribe.
+
+## Properties
+
+This pros and cons list is specified relative to Tock 2.0's syscall ABI.
+
+Pros:
+
+1. No more manually casting everything to a `u32`/`usize` to invoke Command or
+   an upcall; any casting necessary will be handled by the core syscall layer
+   (or code generator).
+1. Type safety: If there is a type mismatch between what userspace expects a
+   syscall to use and what the syscall actually uses, that will be caught
+   (by the kernel for arguments and by the userspace library for return values)
+   and result in an error code rather than data corruption.
+1. Argument count checks: no more `command(..., ..., ..., 0 /* unused */)`;
+   syscall invocations specify exactly the number of arguments the system call
+   needs. The same applies to upcalls.
+1. Command can accept many more arguments and return many more values in both
+   its success and error case.
+
+Cons:
+
+1. Command clobbers many registers. The impact of this is somewhat mitigated on
+   ARM because it has instructions to push and pop multiple registers. It is
+   also somewhat mitigated on RISC-V because it only clobbers caller-saved
+   registers (so the compiler has local context on which registers need to be
+   saved).
+1. The userspace implementation of Yield is now larger, as it has to invoke the
+   upcall itself. On some platforms, this may involve passing arguments via the
+   stack. I would expect Yield to not be inlined with this ABI, whereas in Tock
+   2.0 I would expect it to be inlined.
+1. It adds error paths to the kernel, because system calls (particularly
+   Command) can fail in a new way (incorrect argument type).
+1. Similarly, upcall arguments can error, if the passed `ArbitraryData(4)` does
+   not match the type expected by the userspace driver.
+1. Success versus failure is not defined as part of this ABI. Instead, it is a
+   convention that system calls should define different types for success and
+   failure cases so they can be distinguished (in practice, success types
+   generally won't include `ErrorCode` while failure types will). If an
+   `ArbitraryData` of unexpected type is received, there is no reliable
+   mechanism to detect whether it is a success or error.
```

Comment:
I'm trying to understand if this is important. Return values having a clear set of Success and Failure values was a pretty big part of TRD104. But if this is only when an unexpected type is received, that's sort of irrelevant, as that just means the userspace library implementation is broken.

Reply:
I replaced this point with:

> 1. If an `ArbitraryData` of unexpected type is received, there is no reliable mechanism for userspace to detect whether the system calls succeeded or failed.

With TRD 104, the userspace library could still determine whether the syscall succeeded, even if it doesn't know how to interpret the full return value. Granted, I just checked `libtock-rs`' source and it doesn't appear to support that, so it's probably not very useful.

Reply:
Agreed. The ability to "limp on" with an unexpected return value doesn't seem all that important to me.